### PR TITLE
[SYCL] Put constant initializer list data in non-generic addr space.

### PR DESCRIPTION
**Pull request by kbobrovs**
Big constant initializer lists (>= 16 elements) were represented as a
global in the generic address space. This is incorrect, as this is an
alias AS, and allocation is not possible in it - compiler will have no
clue where to really put data. Gen vector BE crashed on this pattern.
The fix is to put this data into the same address space as a string
literal.

End-to-end test for ESIMD (Gen vector BE) is underway to llvm-test-suite:
https://github.com/intel/llvm-test-suite/pull/91

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

**Comment by bader** [comment]
@kbobrovs, I think https://github.com/intel/llvm/pull/2864 addresses this problem as well. Could you confirm that, please?

If so, I prefer we commit https://github.com/intel/llvm/pull/2864 instead to avoid conflicts between these two patches. 
https://github.com/intel/llvm/pull/2864 implements the approach suggested by clang's CodeGen code owner and significantly reduces the source code difference with llorg.

**Comment by kbobrovs** [comment]
@bader, yes, it seems #2864. Please add a test case from #3005 to #2864 to make sure the scenario we care about works. I will then close #3005.

**Comment by bader** [comment]
> @bader, yes, it seems #2864 solves the problem fixed by #3005. Please add a test case from #3005 to #2864 to make sure the scenario we care about works. I will then close #3005.

I've added similar test to #2864 and it doesn't work for me. The constant array used in initializer list is allocated in generic address space. I'm going to fix this case in scope #2864. Feel free to proceed with this PR, if it's an urgent - fixing regressions caused by #2864 will take some time.

File: clang/test/CodeGenSYCL/local_var_big_init_as.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    // CHECK-NOT: @constinit = private unnamed_addr addrspace(2)
+    // CHECK-NOT: @constinit = private unnamed_addr addrspace(3)
+    // CHECK-NOT: @constinit = private unnamed_addr addrspace(4)
+  } {};
+  int set[32];
+};
+
+__attribute__((sycl_device)) void bar(Test &x);
+
+__attribute__((sycl_device)) void zoo(int *y) {
```
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/local_var_big_init_as.cpp:28]: Can you please explain the need for `zoo `and `vec `in this test?

Comment (kbobrovs) [reply clang/test/CodeGenSYCL/local_var_big_init_as.cpp:28]: `zoo` contains the definition of a vector with static initializer, bar uses it in such a manner that compiler can't replace usage with individual constants. This makes sure `@constinit` global is generated.

**Comment by kbobrovs** [comment]
> Feel free to proceed with this PR, if it's an urgent - fixing regressions caused by #2864 will take some time.

It is pretty urgent - affects important users scenarios. Since this fix is just 3 lines, it should not be hard to replace it with #2864 when it is ready, I believe. So yes, I'd prefer to proceed with this one then.

File: clang/test/CodeGenSYCL/local_var_big_init_as.cpp (hunk 1 above)
Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/local_var_big_init_as.cpp:27]: I'm not sure if `vec` array in `zoo()` makes a difference to `set ` initializer. I don't know if I am missing something but I don't see how it is being used in this test.  I think you can delete the array in `zoo ` and the parameter ` int *y`. I was curious to see the IR without these and so I checked out your patch and ran the test without it. There is no difference in IR. The `@constinit` global is still generated

File: clang/test/CodeGenSYCL/local_var_big_init_as.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+               0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15} {};
+  int set[32];
+};
+// CHECK-DAG: @constinit = private unnamed_addr addrspace(1) constant
+// CHECK: [32 x i32]
+// CHECK: [i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7,
+// CHECK:  i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15,
+// CHECK:  i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7,
+// CHECK:  i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15
+// CHECK  ], align 4
```
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/local_var_big_init_as.cpp:19]: Colon missing

File: clang/test/CodeGenSYCL/local_var_big_init_as.cpp (hunk 1 above)
Comment (kbobrovs) [reply clang/test/CodeGenSYCL/local_var_big_init_as.cpp:27]: Right, `vec` and `y` are redundant. My reply was just about `zoo` basically. Will fix now.

File: clang/test/CodeGenSYCL/local_var_big_init_as.cpp (hunk 2 above)
Comment (kbobrovs) [reply clang/test/CodeGenSYCL/local_var_big_init_as.cpp]: fixed, thanks

**Review by elizabethandrews — APPROVED** [review]
LGTM. Thanks!

**Comment by kbobrovs** [comment]
> End-to-end test for ESIMD (Gen vector BE) is underway to llvm-test-suite.

https://github.com/intel/llvm-test-suite/pull/91

**Comment by kbobrovs** [comment]
@premanandrao - please review/approve as the code owner

**Pull request merged**